lisa/tests/staging/utilclamp.py

```diff
@@ -0,0 +1,296 @@
+        df = df.dropna()
+        df = df[df.active == 1]
+        df['Time'] = df['Time'].cumsum()
+
+        # Frequency to capacity
+        capacities = self._collect_capacities(self.plat_info)
+        for cpu in self.plat_info['nrg-model'].cpu_nodes:
+            cpu_id = cpu.cpus[0]
+            df[cpu_id] = df[cpu_id].map(capacities[cpu_id])
+
+        df['capacity'] = df.apply(lambda x: x[x.cpu], axis=1)
+        df['capacity_max'] = df.apply(
+                lambda x: max(capacities[x.cpu].values()), axis=1)
+
+        return [round(series_mean(df['capacity'], x=df['Time'])),
+                round(series_mean(df['capacity_max'], x=df['Time']))]
+
+    @memoized
+    def _get_analysis(self):
+        task_name = self.rtapp_task_ids_map['task'][0].comm
+
+        df = self.trace.analysis.rta.df_phases(task_name)
+        df = df.reset_index()
+        df.rename(columns={'index': 'start'}, inplace=True)
+        df['end'] = df['start'].shift(-1)
+        df = df[df.phase > 0]
+
+        df[['capacity', 'capacity_max']] = df.apply(
+                lambda x:
+                self._get_task_activation_capacity(task_name, x['start'],
+                                                   x['end']),
+                axis=1, result_type='expand')
+
+        phases = self._get_phases(self.plat_info)
+        phases.insert(0, 0)
+
+        df[['uclamp_val', 'util', 'band']] = df.apply(
+                lambda x: [phases[int(x.phase)][0],
+                           phases[int(x.phase)][1],
+                           phases[int(x.phase)][2]],
+                axis=1, result_type='expand')
+
+        return df
+
+    @classmethod
+    def _test_passed(cls, df, col, minmax_func):
+        df[['expected_min', 'expected_max']] = df.apply(
+                lambda x: minmax_func(x),
+                axis=1, result_type='expand')
+
+        df['pass'] = ((
+            df[col] >= df['expected_min']) &
+            (df[col] <= df['expected_max']))
+
+        print(df)
+
+        return ResultBundle.from_bool(df['pass'].all())
+
+    @FrequencyAnalysis.df_cpus_frequency.used_events
+    @LoadTrackingAnalysis.df_tasks_signal.used_events
+    def test_placement(self) -> ResultBundle:
+        """
+        For each phase, checks if the task placement matches UtilClamp
+        requirements.
+
+        The expected capacity is the one available just above the UtilClamp
+        value +/- 10% (to handle migration time)
+        """
+
+        df = self._get_analysis()
+        max_capacities = sorted(list(dict.fromkeys(
+            self.plat_info['cpu-capacities']['rtapp'].values())))
+
+        def _expected_minmax(row):
+            uclamp_val = row['uclamp_val']
+            expected = 1024
+
+            for cap in max_capacities:
+                # kernel task placement uses 80% margin on the capacity
+                if (cap * 0.8) > uclamp_val:
+                    expected = cap
+                    break
```

Comment:
nit: `for` loops have an optional `break` clause that can be thought as paired with an `if` statement breaking the loop:
```
            for cap in max_capacities:
                # kernel task placement uses 80% margin on the capacity
                if (cap * 0.8) > uclamp_val:
                    expected = cap
                    break
            else:
                expected = 1024
```

It's equivalent to that pseudo-code (there is no binding for `cap`):
```
            if any((cap * 0.8) > uclamp_val for cap in max_capacities):
                expected = cap
            else:
                expected = 1024
```